interp/src/environment.rs

```diff
@@ -24,48 +31,324 @@ impl Environment {
     /// ctx : A context from the IR
     pub fn init(ctx: &ir::RRC<ir::Context>) -> Self {
         Self {
-            map: Environment::construct_map(&ctx.borrow()),
+            //map: Environment::construct_map(&ctx.borrow()),
             context: ctx.clone(),
+            clk: 0,
+            pv_map: Environment::construct_pv_map(&ctx.borrow()),
+            cell_prim_map: Environment::construct_cp_map(&ctx.borrow()),
         }
     }
 
-    /// Returns the value on a port, in a component's cell.
-    // XXX(rachit): Deprecate this method in favor of `get_from_port`
-    pub fn get(&self, component: &ir::Id, cell: &ir::Id, port: &ir::Id) -> u64 {
-        self.map[component][cell][port]
+    pub fn insert(&mut self, port: *const ir::Port, value: Value) {
+        self.pv_map.insert(port, value);
     }
 
-    /// Return the value associated with a component's port.
-    pub fn get_from_port(&self, component: &ir::Id, port: &ir::Port) -> u64 {
-        if port.is_hole() {
-            panic!("Cannot get value from hole")
+    //all of these use parameters as values for constuctors
+    fn construct_cp_map(
+        ctx: &ir::Context,
+    ) -> HashMap<*const ir::Cell, primitives::Primitive> {
+        let mut map = HashMap::new();
+        for comp in &ctx.components {
+            for cell in comp.cells.iter() {
+                let cl: &ir::Cell = &cell.borrow();
+
+                if let ir::CellType::Primitive { name, .. } = &cl.prototype {
```

Comment:
`Cell` has a method called `type_name()` that returns the name of the underlying primitive or the component as an `Option<Id>`. Maybe that would be more appropriate?

Reply:
I think I'd be in favor of leaving it as is since this is specifically for inflating the primitive defns. We'd want different behavior for cells pointing to components (though the interpreter currently doesn't handle this)